#1231 Use polygon centroid for 'point' symbol placements

blanchg opened the pull request:
For Issue #1231. This is a first pass that uses a slightly modified centroid function from turf-centroid.js

It works great for simple polygons, but it doesn't handle holes or complex polygons.

For holes, maybe the solution would be to look at using the formula for [geometric decomposition](https://en.wikipedia.org/wiki/Centroid#By_geometric_decomposition)

For complex polygons (e.g. river polygons ) a more complex straight polygon tree could be used but I haven't found a nicie clean fast solution for that yet.

This is also my first PR for mapbox-gl-js, so please let me know if I have missed something.

mourner commented on js/util/util.js:
```diff
@@ -474,3 +475,27 @@ exports.warnOnce = function(message) {
+        xSum += coord.x - xC;
+        ySum += coord.y - yC;
+        len++;
+    });
```
Please replace with a simple `for` loop for performance and simplicity.

mourner replied:
Also, I think we should use polygon centroid algorithm like this https://en.wikipedia.org/wiki/Centroid#Centroid_of_polygon, which will choose a better center.

blanchg replied:
Excellent thanks for pointing it out.  There was an error in the function on wikipedia that has now been corrected.  I have the additional changes but it causes tests to fail if we keep this change as symbol-placment: 'point'.  

The side effect is that lineStrings will use the centroid which I believe is incorrect.  I did ask if a symbol-placment: 'centroid' should be used in #1231.  This would require changes to the style spec as well, I may need some help/pointers to understand how that is also updated/released.

dpieri commented:
@blanchg thanks for starting this! Lack of this feature has been a huge pain point for us.

In my opinion, [Pole of Inaccessibility](https://en.wikipedia.org/wiki/Pole_of_inaccessibility) is the ideal way to position labels on polygons. I can't speak to the performance implications of doing it in real time compared to a centroid. If they are negligible then I hope you would consider using this method instead.

Here's a Turf [module](https://github.com/Turfjs/turf-polygon-reduce) that implements it, and some more [background](https://github.com/Turfjs/turf/issues/240).

blanchg commented:
That looks like a great solution, however that turf module uses turf-buffer which uses jsts to do the actual bufferring, that would increase the size of mapbox significantly.  I will look to see by how much if we only include the polygon offset/buffer code from jsts.

blanchg commented:
@dpieri I found a small library that performs the Pole Of Inaccessibility and extended it to support holes: https://github.com/blanchg/LabelPoint, however it appears to interfere with a lot of polygon draws so that on higher levels the polygons stop rendering completely and the +/- zoom buttons never get displayed.  Is there a way to find out why this is occuring (is there some sort of timeout maybe?) without going step by step through the code?

dpieri commented:
I'm not sure, maybe @mourner could point you in the right direction.

blanchg commented:
Thanks, I found that the simplification can produce zero area polygons with multiple points that was causing issues.

blanchg commented:
This is ready now.  It continues to use symbol-placment point.

If the ring is closed e.g. first and last point are the same and has an area != 0 then it chooses the Pole of Inaccessibility otherwise it falls back to the first point as it was before.  This works then for points, lines and polygons that have been simplified down to a point/line.

It adds about 1-2kb to the output, fps benchmarks look very similar.

I am positive there are things I can do better so please let me know!  Thanks

mourner commented on js/util/counter_clockwise.js:
```diff
@@ -0,0 +1,6 @@
+// http://bryceboe.com/2006/10/23/line-segment-intersection-algorithm/
+module.exports = function (a, b, c) {
+    return (c.y - a.y) * (b.x - a.x) > (b.y - a.y) * (c.x - a.x);
+};
```
I don't think we need to cut out small functions like this into a separate file. Maybe move to `util.js`? Or maybe we should have `geom_util.js` for all functions like this.

mourner commented on js/util/util.js:
```diff
@@ -454,3 +455,40 @@ exports.warnOnce = function(message) {
+    var f = 6 * area;
+
+    return [cX / f, cY / f];
+};
```
If we use pole of inaccessibility algorithm, this one is no longer used, so should be removed, right?

mollymerp commented:
@blanchg thank you for contributing! 
could you please:
1. rebase your PR to squash your commits and remove commits from other PRs? you can use `git rebase -i origin/master` to do this.
2. add unit tests and/or [render tests](https://github.com/mapbox/mapbox-gl-test-suite)

thanks! 

👓  @lucaswoj @mourner @mollymerp for review

blanchg commented:
Thanks for the reviews @mollymerp and @mourner !

I managed to squish the changes and remove the unrelated ones thanks for pointing that out.  I added some unit tests for polygon_poi and I have a render test but I am uncertain how to proceed to get the render test included.  

I have a feature branch on a fork, I assume I do a PR on the test suite that should only be pulled in once this PR has been?

blanchg commented:
I found that building polygons in the render test suite data have outer rings that don't match the tile specification (2.1).  They have both positive and negative areas as outer rings.  I also found util/classify_rings that handles both cases.  Will update to use that.

blanchg commented:
Updated to use util/classify_rings which now handles polygons and holes of any winding order.  One raster test will fail until it is updated.  I still need some information on how to get the test-suite aligned with this PR.

mollymerp commented:
> I still need some information on how to get the test-suite aligned with this PR.

You can see information on how to update the tests [here](https://github.com/mapbox/mapbox-gl-test-suite#writing-new-tests). Once your tests are satisfactory, make a PR in mapbox/mapbox-gl-test-suite repo and then you can update the sha for that commit in the package.json for this PR. 

I don't see the raster tests failing, I just see the `debug/tile` test failing. Any idea why that might be happening?

@mourner I think you should have final review here.

blanchg commented:
@mollymerp Thanks for the information.  I have that PR submitted now.

The debug/tile test fails because one of the labels is in a closed polygon so it is now rendered at the center of that polygon, every other test doesn't render labels on closed polygons.

I have added the line style and zoomed to the label that moved slightly in the debug/tile test.
![image](https://cloud.githubusercontent.com/assets/86495/16094689/bd9fbe62-3384-11e6-969b-15d06a81f5d3.png)

blanchg commented:
This is now up to date with the head, squashed, unit tests, there is a render test added and updated the failing debug test.  PR in the test-suite and sha pointing to it in package.json.

I think finally it is ready to merge :smiley: :tada: :balloon:

mourner commented:
Fantastic! I'll do a final review tomorrow and merge if everything's good.

dpieri commented:
Any idea when this will make it into a point release?

mourner commented:
This got a bit sidetracked but hopefully we will merge this week, followed by a point release next week.

blanche commented:
i think you ment @blanchg ;)

blanchg commented:
@mourner Thanks for looking at this again!  Updated and squashed.

mourner commented on js/util/polygon_poi.js:
```diff
@@ -0,0 +1,150 @@
+        var r = ((xMax - xMin) * (yMax - yMin));
+        var dx, dy;
+        while (r > precision) {
+            lp = poleScan(xMin, yMin, xMax, yMax, points, holes, allpoints);
```
If precision is > 0, `poleScan` (the expensive part of the algorithm) is always called twice with the same arguments — can we eliminate the redundant call?

mourner commented on js/data/bucket/symbol_bucket.js:
```diff
@@ -332,6 +338,39 @@ SymbolBucket.prototype.addFeature = function(lines, shapedText, shapedIcon, feat
+            holes = polygon.slice(1);
+        }
+        var poi = poleOfInaccessibility(line, holes, 1);
+        anchors = [ new Anchor(poi.x, poi.y, 0) ];
```
Anchor coordinates should probably be rounded, like polygon coordinates.

mourner commented on js/util/polygon_poi.js:
```diff
@@ -0,0 +1,150 @@
+    }
+    var dx = x - xx;
+    var dy = y - yy;
+    // return Math.sqrt(dx * dx + dy * dy);
```
No need for this comment.

mourner commented on js/util/polygon_poi.js:
```diff
@@ -0,0 +1,150 @@
+            c = !c;
+    }
+    return c;
+}
```
Many of the functions introduced here (as well as in `util.js`) are also present in https://github.com/mapbox/mapbox-gl-js/blob/62ac39cf7f1f55a9a82b44e67fb5d71271e97f8b/js/util/intersection_tests.js#L102-L164. Can we eliminate the duplication?

mourner commented on js/util/polygon_poi.js:
```diff
@@ -0,0 +1,150 @@
+        while (r > precision) {
+            lp = poleScan(xMin, yMin, xMax, yMax, points, holes, allpoints);
+            dx = (xMax - xMin) / 24;
+            dy = (yMax - yMin) / 24;
```
Let's move out the magic number `24` into a constant at the top of the file, e.g. `GRID_SIZE`.

mourner commented on js/util/util.js:
```diff
@@ -454,3 +454,60 @@ exports.warnOnce = function(message) {
+    var p2 = points[points.length - 1];
+
+    if (Math.abs(p1.x - p2.x) > 0.001 ||
+        Math.abs(p1.y - p2.y) > 0.001) {
```
Why do we need `0.001` thresholds here if coordinates are expected to be integer?

mourner commented on js/util/polygon_poi.js:
```diff
@@ -0,0 +1,150 @@
+        if (yMax < p.y) yMax = p.y;
+    }
+
+    var allpoints = holes ? points.concat.apply(points, holes) : points;
```
By reorganizing the code a bit, I think we can fully avoid doing `concat` (which copies all the data), gaining a bit of performance.

mourner commented:
Aside from the comments above, I'm also slightly worried about performance — the algorithm used is pretty expensive. E.g. on a sample 4.6k water polygon, it runs for 350ms. Maybe that's fine for an initial implementation, but we should definitely think about a potential follow-up with a faster algorithm.

blanchg commented:
@mourner Thanks for the review, I will address those points.

Can I get that water polygon from somewhere to see if I can improve the performance?  When it is tiled on the map the number of points doesn't seem to get too large though I admit to not having tested on a very complex coastline/waterway.

mourner commented:
@blanchg yeah, unfortunately sometimes polygons can be quite big, especially water, landcover and other natural sources. You can use some of the fixtures from https://github.com/mapbox/earcut/tree/master/test/fixtures — the one I tested is the first ring of `water-huge.json`.

mourner commented:
@blanchg after spending some time on this, I have designed an algorithm (inspired by this one) that is both faster and _guarantees_ global optimum within given precision (unlike the one from the POI paper). On a few test polygons I tried it on, it runs 6-15x faster. I'll push it to NPM as a module and then we can switch to it in this PR.

mourner commented:
Pushed my algorithm to https://github.com/mapbox/polylabel repo. It looks very similar to this one, so the PR could be adapted easily. 

The library I intend to publish handles polygons in GeoJSON format though, so instead of using it as an NPM module, it probably makes sense to bring the code here, replacing `[x, y]` format with the `point-geometry` used in GL JS.

blanchg commented:
That looks very promising! Did you want me to pull that into mapbox or leave it for you?  It might be a few days before I get a chance to work on it.

blanchg commented:
@mourner if this build passes then using your new method has worked and it is ready for another review.

mourner commented on js/util/ajax.js:
```diff
@@ -41,6 +41,7 @@ exports.getJSON = function(url, callback) {
             cache[url] = data;
             callback(null, data);
         } else {
+            console.log("Error", url, response);
```
accidental console.log

mourner commented:
Awesome! Note that additional advantage of this algorithm is that shapes like rectangles get their center as the label point. 

Can you please run the [benchmarks](https://github.com/mapbox/mapbox-gl-js/blob/master/bench/README.md#running-benchmarks) now to make sure we didn't regress anywhere?

mourner commented on js/data/bucket/symbol_bucket.js:
```diff
@@ -333,6 +340,45 @@ SymbolBucket.prototype.addFeature = function(lines, shapedText, shapedIcon, feat
+    }
+
+    if (isPolygon) {
+        var poi = poleOfInaccessibility(polygon, 1);
```
Do you think we need the `1` precision?  It could probably be bumped to something like 16 (which is just 2 pixels).

mourner commented on js/util/polygon_poi.js:
```diff
@@ -0,0 +1,134 @@
+
+/**
+ * Finds an approximation of a polygon's Pole Of Inaccessibiliy https://en.wikipedia.org/wiki/Pole_of_inaccessibility
+ * Note that for perfect rectangles the pole found ISN'T in the center (though the center is a valid POI)
```
The note about perfect rectangles is no longer valid since we get centroid as the first best guess and only choose a different point if it's better in terms of distance. 

I think it's also worth linking to the original repo (https://github.com/mapbox/polylabel) in the comments.

blanchg commented:
@mourner Thanks very much for the review I have updated as needed.

# Benchmarks Master/HEAD

**buffer:** 1,435 ms
**fps:** 60 fps
**frame-duration:** 8.4 ms, 3% > 16ms
**query-point:** 1.12 ms
**query-box:** 72.96 ms
**geojson-setdata-small:** 14 ms
**geojson-setdata-large:** 206 ms

# Benchmarks this PR

**buffer:** 1,320 ms
**fps:** 59 fps
**frame-duration:** 8.7 ms, 3% > 16ms
**query-point:** 1.03 ms
**query-box:** 76.07 ms
**geojson-setdata-small:** 13 ms
**geojson-setdata-large:** 210 ms

mourner commented on js/data/bucket/symbol_bucket.js:
```diff
@@ -333,6 +340,46 @@ SymbolBucket.prototype.addFeature = function(lines, shapedText, shapedIcon, feat
+    if (isPolygon) {
+        // 16 here represents 2 pixels
+        var poi = poleOfInaccessibility(polygon, 16);
+        if (poi === null) {
```
I don't think this can ever be null. The algorithm always returns some point.

mourner commented on js/data/bucket/symbol_bucket.js:
```diff
@@ -333,6 +340,46 @@ SymbolBucket.prototype.addFeature = function(lines, shapedText, shapedIcon, feat
+                isHole = line === polygon[i];
+            }
+        }
+    }
```
It looks like for each polygon ring, you go through all other rings doing equality comparisons, which is O(n^2). However it does not seem to be necessary — you could instead add symbol instances as you loop through the rings above. This would need some code reorganization above.

mourner commented on js/util/polygon_poi.js:
```diff
@@ -0,0 +1,134 @@
+    if (polygons.length === 0) {
+        return null;
+    }
+    if (!isClosedPolygon(polygons[0])) {
```
This check should probably live outside of the poi function.

mourner commented on js/util/polygon_poi.js:
```diff
@@ -0,0 +1,134 @@
+ *
+ * @returns {Point} Pole of Inaccessibiliy.
+ */
+module.exports = function (polygons, precision, debug) {
```
`polygons` is a bit confusing here because it really means rings of a single polygon in this case. Perhaps `rings` would be better?

blanchg commented:
@mourner all excellent points, hopefully I addressed them.

# Benchmarks Master/HEAD

**buffer:** 1,435 ms
**fps:** 60 fps
**frame-duration:** 8.4 ms, 3% > 16ms
**query-point:** 1.12 ms
**query-box:** 72.96 ms
**geojson-setdata-small:** 14 ms
**geojson-setdata-large:** 206 ms
# Benchmarks PR

**buffer:** 1,350 ms
**fps:** 59 fps
**frame-duration:** 8.3 ms, 4% > 16ms
**query-point:** 1.00 ms
**query-box:** 72.39 ms
**geojson-setdata-small:** 15 ms
**geojson-setdata-large:** 204 ms

mourner commented:
This looks pretty good now, thanks for addressing the feedback! I'll let @lucaswoj take a quick final look just in case, and aim to merge soon.

lucaswoj commented on js/data/bucket/symbol_bucket.js:
```diff
@@ -14,6 +14,8 @@ var clipLine = require('../../symbol/clip_line');
 var util = require('../../util/util');
 var loadGeometry = require('../load_geometry');
 var CollisionFeature = require('../../symbol/collision_feature');
+var poleOfInaccessibility = require('../../util/polygon_poi');
```
An (admittedly unwritten) convention in GL JS is that file names match the names of the classes / functions they export. Can we rename `polygon_poi` -> `find_pole_of_inaccessibility` or similar?

lucaswoj commented:
This looks good to me! I'm excited to 🚢 .

blanchg commented:
@mourner @mollymerp @lucaswoj Thank you all for your help, I think this is finally ready :D :+1:

mourner commented:
Set up a JSBin debug page to demonstrate how it works in practice, using the choropleth from examples as a base but adding a population label: http://output.jsbin.com/tajiwociha

Generally it's pretty cool, but there are a few issues that are worth noting, although they don't necessarily warrant an immediate fix, and not all are possible to fix (e.g. because of tiling or the the way algorithm works).

 - You have to use `"symbol-avoid-edges": true`, otherwise there will be collisions across tile boundaries because the neighboring points don't match (they represent different parts cut off from a polygon by tiling).
- Due to buffers around polygons clipped by tile boundaries, labels in such polygons are placed closer to the boundary than needed; the polygons should probably be clipped before running the algorithm for more correct results.
- Polygons split across several tiles will of course have several labels.
- Label positions sometimes "jump" between zoom levels, because the algorithm can pick a different point in a higher-precision/bigger coords version of the same polygon; this is exaggerated on this particular example because of many rectangle-like shapes.

@samanpwbb can you play with the sample and tell us what you think? Should we go ahead with the feature in its current state?

samanpwbb commented:
It's much better than what we have now, so I believe that it is worth shipping in it's current state.

> You have to use "symbol-avoid-edges": true

That's fine – i know this is a hard / impossible problem to fix. Mapnik was similar.

> Due to buffers around polygons clipped by tile boundaries, labels in such polygons are placed closer to the boundary than needed

This problem is noticeable – definitely worth following up on: 
![screen shot 2016-08-18 at 10 35 24 am](https://cloud.githubusercontent.com/assets/108094/17777710/89c9f4cc-652f-11e6-991e-73a9ffcbddc7.png)

> Polygons split across several tiles will of course have several labels.

That's fine, Mapnik had the same issue. vector tiles ¯\_(ツ)_/¯

> Label positions sometimes "jump" between zoom level

Not a big deal.

I'm also seeing an issue with labels on features I expect to be smaller parts of multipolygons. Is this just a data issue, or is mapbox gl unable to distinguish multipolygons?

![screen shot 2016-08-18 at 10 29 06 am](https://cloud.githubusercontent.com/assets/108094/17777788/c5aa03b0-652f-11e6-8d43-0463e08fb63c.png)

mourner commented:
> I'm also seeing an issue with labels on features I expect to be smaller parts of multipolygons. Is this just a data issue, or is mapbox gl unable to distinguish multipolygons?

It currently just adds a label for every polygon in a multipolygon. We could alter the algororithm to filter out labels for polygon where anchor point has a distance to the polygon below a certain threshold, but this would probably require a new obscure style-spec property, or have a very good default. Another option is to always only add one label to a multipolygon, choosing an anchor in one of the polygons that fits the best.

mourner commented:
It's also worth noting that the performance cost on the worker side is pretty high — on this example but without a basemap (just leaving the population map alone), placing polygon anchors takes ~18% of the worker time. If I switch from 2px to 4px precision, it takes ~13%. It may be acceptable though, since this is an opt-in feature, and doesn't affect our styles — only custom visualizations where you explicitly want to use auto polygon anchors.

samanpwbb commented:
> Another option is to always only add one label to a multipolygon, choosing an anchor in one of the polygons that fits the best.

I can see cases where both would be desireable, but _most of the time_ a multipolygon should only get a single label. you'd almost never want to include more than one label on a choropleth data viz (and this could even result in a map that 'lies' – making it look like a place has double the value that it actually has). 

On the other hand, a cartographer may want to place a tree icon in the center of each piece of a park multipolygon. 

I vote to enforce 1 label per multipolygon, leaving open the option of adding a new property to give users control in the future.

mourner commented:
@samanpwbb makes sense! Regarding polygons that have labels placed close to boundaries, I referred to a different case:

![image](https://cloud.githubusercontent.com/assets/25395/17778589/3924b4ba-656d-11e6-8232-7ec90e2baab0.png)

Here, the label is not further up because it calculates anchor based on the polygon geometry with a buffer, including an "invisible" part on the bottom. 

And regarding the case below, — it's just a property of the algorithm — in this case a label in the center may have a few pixels smaller distance to the polygon, and considered worse than the one picked; to fix this, we would have to modify the algorithm to somehow prefer points closer to centroid even if they have smaller distance, but this is not trivial.

![image](https://cloud.githubusercontent.com/assets/25395/17778668/83e3e796-656d-11e6-8f2c-7c3803d15a9e.png)

samanpwbb commented:
> We could alter the algororithm to filter out labels for polygon where anchor point has a distance to the polygon below a certain threshold

After thinking more about the problem and various edge cases, this is my favorite option for handling multi-poly labelling. It wouldn't add a new property and would ideally look good in most cases. I also recognize that this is the hard approach. Again, like I said in the original post, my take on this PR is that it's a major improvement over what we have now no matter what, and I think we should merge something imperfect sooner (and ticket follow up work) rather than wait until it's perfect.

mourner commented:
> It wouldn't add a new property and would ideally look good in most cases.

How do we pick the threshold then? E.g. something arbitrary like 2 "ems" assuming em = 24px?

samanpwbb commented:
> How do we pick the threshold then? E.g. something arbitrary like 2 "ems" assuming em = 24px?

Yep something like that sounds good. See how it feels. 

--- 

Let me bring up another idea. I don't know if it's at all technically feasible to incorporate into your label placement algorithm. Archipelago-type features are really common in the world. Louisiana is mostly a single mass, but has an archipelago on it's southern tip. But what about a Phillipines?

Is there a way to incorporate multipolygons into the label placement algorithm in such a way that the label still ends up getting placed inside a shape, so we are able to handle archipelagos well?

blanchg commented:
I found if you ignore polygons that have a cellSize < precision * 2 it seems to handle the US quite well and fixes the label showing up on the islands off Louisiana and Texas when you are zoomed out.  This could cause issues for archipelagos though as it won't show any labels until one of the islands is a few pixels big.  Seems to speed up the rendering too.

http://jsbin.com/lopilikohe/1

`
var cellSize = Math.min(width, height);

if (cellSize < precision * 10) {
    return null;
}
`

fedex1 commented:
Here's an example that I'm using to test.

Before centroid change: https://city.tidalforce.org/nycschoolallocation_usm.html
See the "white label"  "District 16"
![screen shot 2016-08-18 at 11 08 43 pm](https://cloud.githubusercontent.com/assets/167614/17797722/2c58a1fc-6599-11e6-873d-36ced9e7d969.png)

With the centroid change (I cloned mapbox-gl-js and switched to the pull request branch and built)
![screen shot 2016-08-18 at 11 07 45 pm](https://cloud.githubusercontent.com/assets/167614/17797712/13a3904a-6599-11e6-8891-480f5162d009.png)

It is definitely an improvement.

A few questions:

- Is there supposed to be multiple labels for a small polygon?  this may be a result of my data so I apologize if this is a data issue.

- I noticed that the copyright style change
```

 map.style.sources['mapbox://cityregister.nyc_school_allocation_titlei'].attribution = [
                '&copy;BrooklynMarathon',
                '<a href="http://schools.nyc.gov/offices/d_chanc_oper/budget/dbor/allocationmemo/fy16_17/FY17_PDF/sam08_T2.xls">City of New York, Department of Education</a>'
            ].join(' | ');

```
no longer seems to work.

as it did in

```
<script src='https://api.tiles.mapbox.com/mapbox-gl-js/v0.19.1/mapbox-gl.js'></script>
<link href='https://api.mapbox.com/mapbox-gl-js/v0.19.1/mapbox-gl.css' rel='stylesheet' />


```

blanchg commented:
@fedex1 I don't think this PR has any changes that impact the copyright maybe the others can comment on this
Those multiple labels are likely due to the tiles if you turn on the tile debug lines it should highlight this.

:+1: for giving it a go and demonstrating feedback

fedex1 commented:
@blanchg Yes that is correct.  Thanks for the reply!

with debug lines on. https://city.tidalforce.org/nycschoolallocation_usm_centroid.html

![screen shot 2016-08-19 at 12 43 01 am](https://cloud.githubusercontent.com/assets/167614/17799223/b09bee1c-65a6-11e6-966c-2f54313a8f63.png)

lucaswoj commented:
Rebased and squashed 👉 https://github.com/mapbox/mapbox-gl-js/pull/3038

anna-geo commented:
Could you please explain if it is possible now to show only one text label in polygon center even if it`s splited into several tiles? I would like to avoid situation like on the last image of @fedex1 where label "District 16" is repeated several times inside one district.

mourner commented:
@anna-geo currently not possible unfortunately, and a very hard issue to address.

lucaswoj commented:
@anna-geo I suggest that you do that processing during tile generation time and produce point geometries where the label should go.

anna-geo commented:
@mourner @lucaswoj Thank you for reply! Will be waiting for this feature.
